File: src/julia/01.jl
const INPUT_PATH = ARGS[1]

using Test

# Part I
function getnumber(line)
    i = findfirst(isdigit, line)
    j = findlast(isdigit, line)
    return parse(Int, string(line[i],line[j]))
end

function part1()
    lines = readlines(INPUT_PATH)
    println(sum(getnumber, lines))
end

part1() # Answer: 55123

const lookup = Dict("one" => 1, "two" => 2, "three" => 3, "four" => 4, "five" => 5, "six" => 6, "seven" => 7, "eight" => 8, "nine" => 9)

const RE = Regex("$(join(keys(lookup), '|'))|\\d")

# Part II
function get_num_from_string(str)
	dig = tryparse(Int, str)
    !isnothing(dig) && return dig
    return lookup[str]
end

function getnumber_part2(line)
    captures = collect(eachmatch(RE, line; overlap=true))
    num_first = get_num_from_string(captures[begin].match)
    num_last = get_num_from_string(captures[end].match)
    return num_first * 10 + num_last
end

# @testset "getnumber_part2" begin
#     @test getnumber_part2("two1nine") == 29
#     @test getnumber_part2("6vzpjhtwonemc") == 61
#     @test getnumber_part2("eightwothree") == 83
#     @test getnumber_part2("abcone2threexyz") == 13
#     @test getnumber_part2("xtwone3four") == 24
#     @test getnumber_part2("4nineeightseven2") == 42
#     @test getnumber_part2("zoneight234") == 14
#     @test getnumber_part2("7pqrstsixteen") == 76
# end

function part2()
    lines = readlines(INPUT_PATH)
    println(sum(getnumber_part2, lines))
end

part2() # Answer: 55260

File: src/cpp/01.cpp
#include <iostream>
#include <fstream>

auto strtonum_part1(std::string str)
{
    auto result = 0;
    auto num_first = -1;
    auto num_last = -1;
    for (int i = 0; i < str.length(); i++)
    {
        if (std::isdigit(str[i]))
        {
            if (num_first == -1)
            {
                num_first = str[i] - '0';
            }
            num_last = str[i] - '0';
        }
    }
    result += num_first * 10 + num_last;
    return result;
}

auto substrtonum(std::string str)
{
    if (str.starts_with("one"))
        return 1;
    else if (str.starts_with("two"))
        return 2;
    else if (str.starts_with("three"))
        return 3;
    else if (str.starts_with("four"))
        return 4;
    else if (str.starts_with("five"))
        return 5;
    else if (str.starts_with("six"))
        return 6;
    else if (str.starts_with("seven"))
        return 7;
    else if (str.starts_with("eight"))
        return 8;
    else if (str.starts_with("nine"))
        return 9;
    else
        return -1;
}

auto strtonum_part2(std::string str)
{
    auto result = 0;
    auto num_first = -1;
    auto num_last = -1;
    for (int i = 0; i < str.length(); i++)
    {
        auto num = -1;
        if (std::isdigit(str[i]))
            num = str[i] - '0';
        else
        {
            num = substrtonum(str.substr(i));
            if (num == -1)
                continue;
        }

        if (num_first == -1)
        {
            num_first = num;
        }
        num_last = num;
    }
    result += num_first * 10 + num_last;
    return result;
}

int main()
{
    std::ifstream file("inputs/01_bauerc.txt");
    std::string str;
    auto result_part1 = 0;
    auto result_part2 = 0;
    while (std::getline(file, str))
    {
        // std::cout << str << std::endl;
        result_part1 += strtonum_part1(str);
        result_part2 += strtonum_part2(str);
    }
    std::cout << "Answer: " << result_part1 << std::endl;
    std::cout << "Answer: " << result_part2 << std::endl;
    return 0;
}
